lib/ansible/module_utils/vmware.py

```diff
@@ -828,41 +831,190 @@ def is_vcenter(self):
+
+                # Nested folder does not return trailing /
+                if not dcpath.endswith('/'):
+                    dcpath += '/'
+
+                if user_folder in [None, '', '/']:
+                    # User provided blank value or
+                    # User provided only root value, we fail
+                    self.module.fail_json(msg="vmware_guest found multiple virtual machines with same "
+                                              "name [%s], please specify folder path other than blank "
+                                              "or '/'" % self.params['name'])
+                elif user_folder == '/vm' or user_folder.startswith('/vm/'):
+                    # User provided VMware default vm folder with / or
+                    # User provided nested folder under VMware default vm folder i.e. folder = /vm/india/finance
+                    user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)
+                elif user_folder == 'vm':
+                    # User provided VMware default vm folder
+                    user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)
+                elif user_folder.startswith("%s%s/vm" % (dcpath, user_defined_dc)) or \
+                        user_folder.startswith("/%s" % user_defined_dc):
+                    # User defined datacenter is not nested i.e. dcpath = '/' , or
+                    # User defined datacenter is nested i.e. dcpath = '/F0/DC0' or
+                    # User provided folder starts with / and datacenter i.e. folder = /ha-datacenter/
+                    user_desired_path = user_folder
+                elif user_folder.startswith('/'):
+                    # User defined folder is under VMware default vm folder i.e. folder = /india/finance
+                    user_desired_path = "%s%s/vm%s" % (dcpath, user_defined_dc, user_folder)
+                elif user_folder.startswith("%s/vm" % user_defined_dc):
+                    # User defined folder starts with datacenter without '/' i.e.
+                    # folder = DC0/vm/india/finance or
+                    # folder = DC0/vm
+                    user_desired_path = user_folder
+                elif user_folder == '%s' % user_defined_dc:
+                    # User defined folder is datacenter i.e. folder = 'DC0'
+                    user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)
+                else:
+                    # get_vm is not super cow !!!
+                    user_desired_path = user_folder
+
+                for vm in vms:
+                    # Check if user has provided same path as virtual machine
+                    actual_vm_folder_path = self.get_vm_path(content=self.content, vm_name=vm)
+                    if not actual_vm_folder_path.startswith("%s%s" % (dcpath, user_defined_dc)):
+                        continue
+                    if user_desired_path in actual_vm_folder_path:
```

Comment:
I still think this could be simplify by:
```
user_folder = self.params['folder']
if not user_folder.startswith('/'):
    user_folder = '/' + user_folder
if not user_folder.endswith('/'):
    user_folder += '/'

…

for vm in vms:
    …
    if user_path in actual_vm_folder_path:
        vm_obj = vm
        break
```

But if you are sure about the user_desired_path computation, that's OK for me.

Reply:
Yes. I would like to keep user_folder unaltered.

Reply:
Then:
```
user_folder = self.params['folder']
slashed_user_folder = user_folder
if not slashed_user_folder.startswith('/'):
    slashed_user_folder = '/' + slashed_user_folder
if not slashed_user_folder.endswith('/'):
    slashed_user_folder += '/'

…

for vm in vms:
    …
    if slashed_user_folder in actual_vm_folder_path:
        vm_obj = vm
        break
```

My point/remark/question wasn't about altering `user_folder` but about the `user_desired_path` computation and usage.

The important part is that I suggest to replace:
```
if user_desired_path in actual_vm_folder_path:
```
by:
```
if slashed_user_folder in actual_vm_folder_path:
```

Again if you are sure about the user_desired_path computation, that's OK for me.

Reply:
I am sure about user_desired_path computation. So I think, I will keep it as it is.

For example, `user_folder` is '/DC0/vm' and `actual_vm_folder_path` for two VMs with same name is '/DC0/vm/finance' and '/DC0/vm/marketing' and user wants VM with '/DC0/vm/finance' path. So there are chances that user might end up in getting wrong VM if we just use `user_folder`. I think computation of user_desired_path might decrease this chance. 

Let me know if it make sense for you.

Reply:
@jctanner What do you think on this situation ?

Reply:
> For example, user_folder is '/DC0/vm' and actual_vm_folder_path for two VMs with same name is '/DC0/vm/finance' and '/DC0/vm/marketing' and user wants VM with '/DC0/vm/finance' path. So there are chances that user might end up in getting wrong VM if we just use user_folder. I think computation of user_desired_path might decrease this chance.

How can you only get the VM in '/DC0/vm/finance' if `user_folder` is '/DC0/vm' ?

If `user_folder` is `/DC0/vm` we fall into the `user_folder.startswith("%s%s/vm" % (dcpath, user_defined_dc))` condition and [user_desired_path = user_folder](https://github.com/ansible/ansible/pull/33525/files#diff-d5ca044435f1fc628995b038d2e238f7R949).
So you get both VM from '/DC0/vm/finance' and '/DC0/vm/marketing'.

################

If you mean to write `/DC0/vm/finance`:
If `user_folder` is `/DC0/vm/finance`, we still fall into the `user_folder.startswith("%s%s/vm" % (dcpath, user_defined_dc))` condition and [user_desired_path = user_folder](https://github.com/ansible/ansible/pull/33525/files#diff-d5ca044435f1fc628995b038d2e238f7R949). So using `user_desired_path` or `user_folder` is exactly the same…

I don't get your example.

Reply:
@Akasurde 
```
+                if user_folder in [None, '', '/']:
+                    # User provided blank value or
+                    # User provided only root value, we fail
+                    self.module.fail_json(msg="vmware_guest found multiple virtual machines with same "
+                                              "name [%s], please specify folder path other than blank "
+                                              "or '/'" % self.params['name'])
+                elif user_folder == '/vm' or user_folder.startswith('/vm/'):
+                    # User provided VMware default vm folder with / or
+                    # User provided nested folder under VMware default vm folder i.e. folder = /vm/india/finance
+                    user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)
+                elif user_folder == 'vm' or user_folder == '%s' % user_defined_dc:
+                    # User provided VMware default vm folder
+                    # User defined folder is datacenter i.e. folder = 'DC0'
+                    user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)
+                else:
+                    # User defined datacenter is not nested i.e. dcpath = '/' , or
+                    # User defined datacenter is nested i.e. dcpath = '/F0/DC0' or
+                    # User provided folder starts with / and datacenter i.e. folder = /ha-datacenter/ or
+                    # User defined folder starts with datacenter without '/' i.e.
+                    # folder = DC0/vm/india/finance or
+                    # folder = DC0/vm
+                    user_desired_path = user_folder
+
+                for vm in vms:
+                    # Check if user has provided same path as virtual machine
+                    actual_vm_folder_path = self.get_vm_path(content=self.content, vm_name=vm)
+                    if not actual_vm_folder_path.startswith("%s%s" % (dcpath, user_defined_dc)):
+                        continue
+                    if user_desired_path in actual_vm_folder_path:
+                        vm_obj = vm
+                        break
```

# `if user_folder in [None, '', '/']` → fail
OK, I am good with that.

# `elif user_folder == '/vm' …` → `user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)`

If user_folder == '/vm', user_desired_path will be = to "/dcpath/user_defined_dc/vm". So if  `actual_vm_folder_path.startswith("%s%s" % (dcpath, user_defined_dc))`, user_desired_path will always be in actual_vm_folder_path.

The case `user_folder == '/vm'` is useless since we get the same result with user_desired_path = user_folder.

# `or user_folder.startswith('/vm/'):` → `user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)`

It bring a bit of precision (compared to user_desired_path = user_folder), so Ok for me

# `elif user_folder == 'vm' …` → `user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)`

Same as before, the case `user_folder == 'vm'` is useless since we get the same result with user_desired_path = user_folder.

# `or user_folder == '%s' % user_defined_dc:` → `user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)`

Side note: `user_folder == '%s' % user_defined_dc` should be written `user_folder == user_defined_dc`.

Same as before, the case `user_folder == user_defined_dc` is useless since we get the same result with user_desired_path = user_folder.